### PR TITLE
Plan 17 + implementation: readable filenames (title-derived slugs, frontmatter id, rename-on-settle)

### PR DESCRIPTION
**Pull request by maccman**
## Summary

Plan 17 finishes the note-identity contract Plan 02 specified but the first wave shortcut: regular notes get **title-derived filenames** (`notes/meeting-notes.md`, not `notes/01kts4w0….md`), a **frontmatter `id`** (ULID) that is both written *and load-bearing*, and **file renames that follow settled title changes**.

Numbered 17 because #109 owns Plan 16.

## Key decisions (recorded in the plan)

- **Pure lowercase slug, `-2` suffix only on collision**, capped at a single 60 code points (covers both readability and the 255-byte basename limit) — lowercase-only makes APFS/NTFS case-insensitivity and git case-sensitivity agree by construction. Golden-corpus tests freeze the slug rules.
- **Frontmatter `id:` is real identity** — written at creation, indexed (schema v7), and consumed: the reconcile and the live watcher **heal external renames by id** (`index_move` migrates projection rows when a vanished row and an appeared file share an id), so Finder/Obsidian/sync-pull renames don't drop embeddings or derived state — and the open pane, route, and history follow. Ambiguous ids never pair; **sync forks surface in Settings → Backup**.
- **An occupied move destination always refuses** — db row or disk file, one rule at both layers; nothing is ever deleted or overwritten, and a refused rename costs only cosmetic filename drift retried on the next settle. (Replaced an earlier same-id adoption heuristic — the race window is sub-millisecond behind a 5s save-quiet gate.)
- **A rename is never a re-embed** — both move paths carry `embedding_chunks` rows (BYOK re-embeds cost the user money). Pinned state and conflict flags survive too.
- **No migration** — pre-launch, there are no pre-Plan-17 graphs to convert; ULID-born notes adopt their slug through the birth/rename path when first titled.
- Wiki links store titles, not paths, so file renames break zero links — the property that makes this plan cheap.

## Implementation

- [x] **Slugs at birth**: `slugForTitle` (golden-corpus frozen) + collision picker (index + disk via `note_exists`, one shared probe loop); `createNoteWithTitle` writes `notes/<slug>.md` with `id:`; ⌘N seeds carry `id:`; schema v7 partial index on `notes.id`
- [x] **Filenames follow titles**: `note_move_indexed` (rows commit first, then the fs rename; a failed rename compensates with a reverse row-move); births (`from: null`) and renames settle through the 07b quiet timer; live sessions retarget in place; the router rewrites history entries (test-pinned); the pane **adopts** the retargeted session — same `sessionEpoch`, the editor keeps its cursor
- [x] **Document lifecycle as an explicit protocol**: create/adopt/teardown/hand-off lives in the unit-tested `document-binding` controller; `useNoteDocument` is a thin React adapter; watcher reconciliation reads the session's current path (never a stale prop)
- [x] **Id-based move healing**: `index_move` + pure `pairMovesById` (ambiguity-refusing) wired into `reconcileIndex` and the watcher batch path via `detectExternalMoves`, with `onMoved` → `followHealedMove` carrying sessions/routes; Reflect's own move echo never pairs by construction
- [x] **Sync rename matrix** (Rust, two-device fixtures): rename+edit converges via libgit2 rename detection; add/add surfaces as marker conflicts; rename/rename keeps both files, never wedges — and surfaces as a duplicate-id fork in settings
- [x] **Simplification pass**: removed the racing-save adoption complex, the dual slug cap, and the ULID→slug migration (~1,200 lines of edge-case and backwards-compat machinery) — pre-launch, no compatibility to keep

## Verification

- `pnpm check` clean
- TS: 609 desktop + 498 core tests pass
- Rust: 91 `reflect-open` + 18 `reflect-cli` tests pass (move txn, echo-survival, merge matrix)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **High Risk**
> Touches core note identity, transactional index/fs moves, rename orchestration, and sync merge behavior—failures could fork notes, lose embeddings, or leave index and disk diverged.
> 
> **Overview**
> Plan 17 delivers **readable note filenames** tied to titles: new notes use slug paths with **`id:`** frontmatter (schema v7), and **settled title changes** run link rewrite, old-title alias, then a coordinated **file + index move** that keeps embeddings and derived state.
> 
> **Backend:** `note_move_indexed` commits projection row moves first, then renames on disk (with compensation on failure); `index_move` heals external renames by id. The old fs-only `note_move` is removed in favor of `note_exists` for collision checks. `move_note` migrates all path-keyed tables including embedding chunks.
> 
> **Desktop:** The rename coordinator gains a move phase and birth handling (`from: null`); `moveNoteCarryingSession`, `document-binding` (adopt retargeted sessions), `sessionEpoch` editor keys, and `onNoteMoved` / router history rewrites keep the editor and navigation on the new path. Settings shows **duplicate-id sync forks**.
> 
> **Core:** `slugForTitle`, `availableNotePath` / `slugPathForTitle`, and id-based **move healing** in reconcile and live watcher batches via `onMoved`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 68ac874d2ddd451cd842e596ceaecdd639ae4c2a. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Human-readable title-derived filenames with collision handling and durable frontmatter IDs.
  * Safe, transactional note moves that preserve embeddings, editor sessions, and history; external renames are detected and healed.
  * Router and open-editor sessions automatically retarget on moves; global note-moved events notify subscribers.
  * Untitled-note seed and first-title “birth” behavior.

* **Improvements**
  * Improved duplicate-ID detection and sync reconciliation.
  * Clearer operation failure messaging surfaced in the UI.

* **Documentation**
  * Added Plan 17 and readable-filenames docs.

### CODE REVIEW

File: apps/desktop/src-tauri/src/db/mod.rs [hunk 1]
```diff
@@ -144,6 +144,38 @@ pub fn index_remove(path: String, generation: u64, index: State<IndexState>) ->
...
+    let root = crate::fs::root_for_generation(&graph, generation)?;
+    let mut state = lock_state(&index)?;
+    let conn = state.conn.as_mut().ok_or_else(AppError::no_graph)?;
+    let tx = conn.transaction()?;
+    // Child tables FK `notes(path)`; deferring lets the parent key move first
+    // and the constraint re-check at commit, when the children have followed.
+    tx.execute_batch("PRAGMA defer_foreign_keys = ON;")?;
+    write::move_note(&tx, &from, &to)?;
+    crate::fs::move_note_file(&root, &from, &to)?;
+    tx.commit()?;
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src-tauri/src/db/mod.rs:202]: Fixed in eefa66d: the row-move transaction now commits before the fs rename, so a failed commit touches nothing — and a failed rename runs a compensating reverse row-move (best-effort; the id-based reconcile heals any residue). The rename can no longer sit inside the transaction.

File: apps/desktop/src-tauri/src/fs/mod.rs [hunk 2]
```diff
@@ -215,6 +215,36 @@ pub fn asset_write(
...
+/// the rename pipeline (Plan 17) the only realistic cause is the note's own
+/// just-retargeted save landing at the destination first — that file carries
+/// the *newest* content, and overwriting it with the stale source would lose
+/// the very keystrokes the save persisted.
+pub(crate) fn move_note_file(root: &Path, from: &str, to: &str) -> AppResult<()> {
+    let from_abs = resolve(root, from)?;
+    let to_abs = resolve(root, to)?;
+    if to_abs.is_file() {
+        fs::remove_file(from_abs)?;
+        return Ok(());
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src-tauri/src/fs/mod.rs]: Fixed in eefa66d: an occupied destination is consumed only when both files carry the same frontmatter id (the note's own racing save, whose bytes are newest); a foreign file now refuses the move — the compensation puts the rows back and nothing is deleted. Pinned by four fs-level tests.

File: apps/desktop/src/editor/move-note.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    await owner.flush()
+    owner.retarget(to)
+    retargetOpenDocument(from, to)
+  }
+  try {
+    await moveNoteIndexed(from, to, generation)
+  } catch (cause) {
+    owner?.retarget(from)
+    retargetOpenDocument(to, from)
+    throw cause
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/move-note.ts:45]: Fixed in 24b4133: retargetOpenDocument is now identity-guarded (moves an entry only when it holds the given session) and the compensation runs only when a session was actually carried — a foreign pane's document at the destination stays put. Pinned by a regression test.

File: apps/desktop/src-tauri/src/db/write.rs [hunk 4]
```diff
@@ -137,6 +137,49 @@ pub(super) fn apply_note(conn: &Connection, note: &IndexedNote) -> AppResult<()>
...
+/// at the new path.
+pub(super) fn move_note(conn: &Connection, from: &str, to: &str) -> AppResult<()> {
+    let occupied: bool = conn
+        .prepare_cached("SELECT 1 FROM notes WHERE path = ?1")?
+        .exists(params![to])?;
+    if occupied {
+        return Err(crate::error::AppError::io(format!(
+            "cannot move note: {to} is already indexed"
+        )));
+    }
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src-tauri/src/db/write.rs:164]: Fixed in 8f5c9c7: move_note now resolves the same-id race exactly like move_note_file — the fresher destination row (the note's own racing save) is adopted, the stale source row drops, and its embedding chunks carry over; different/missing ids still refuse. Pinned by three db tests.

File: apps/desktop/src/lib/filename-migration.ts [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        : ''
+    operation.fail(
+      `renamed ${result.moved} of ${count} — ${result.failed.length} failed and keep their old filenames${skippedToo}; reopening the graph offers the rest again`,
+    )
+  } else {
+    // Pure skips need no alarm: the scan excludes open/conflicted notes up
+    // front, so a skip here is a mid-run race, and the next graph open
+    // simply offers the remainder.
+    operation.done()
+  }
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/lib/filename-migration.ts]: Fixed in c9f2bd9: the operations store gains a notice(message) terminal state (completes, lingers like a failure, styled neutrally) and runFilenameMigration reports 'renamed X of N — M skipped' through it whenever the count falls short without failures. Pinned by store + lib tests.

File: apps/desktop/src/components/filename-migration-prompt.tsx [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    }))
+  }
+
+  const accept = (): void => {
+    setDismissed(true)
+    void runFilenameMigration({ candidates, generation })
+  }
+
+  return (
+    <Dialog open onOpenChange={(open) => !open && setDismissed(true)}>
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/components/filename-migration-prompt.tsx]: Working as designed, now documented + test-pinned (f9369c7): Esc/outside-click is a deliberate defer — a reflexive startup Esc must not permanently kill the offer, so nothing is recorded and the next graph open re-offers; only the explicit 'Keep current names' button is sticky. Filenames remain fully functional meanwhile (new notes get slugs regardless).

File: apps/desktop/src-tauri/src/db/write.rs [hunk 7]
```diff
@@ -137,6 +137,82 @@ pub(super) fn apply_note(conn: &Connection, note: &IndexedNote) -> AppResult<()>
...
+            // Both sides embedded already (rare — the race window is small):
+            // the destination's chunks win; drop the source's so no ghost
+            // rows linger at a path without a note.
+            super::embed_write::remove_chunks(conn, from)?;
+        } else {
+            conn.prepare_cached("UPDATE embedding_chunks SET note_path = ?2 WHERE note_path = ?1")?
+                .execute(params![from, to])?;
+        }
+        remove_note(conn, from)?;
+        return Ok(());
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src-tauri/src/db/write.rs]: Working as designed, now documented + test-pinned (339115c): there are no index-only fields on the notes row — is_pinned/pinned_order/has_conflict are all content-derived projections (migrations 0004/0006), and the destination row is the same note's *fresher* projection (the racing save). Merging the stale source's flags would resurrect exactly the changes that save just made (e.g. re-pin a just-unpinned note). The adopt test now pins the freshness-wins direction; embedding chunks (the one expensive path-keyed derivative) already carry over.

File: apps/desktop/src/lib/note-moves.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+ * landed).
+ */
+export function emitNoteMoved(from: string, to: string): void {
+  for (const listener of [...listeners]) {
+    try {
+      listener(from, to)
+    } catch (cause) {
+      console.error('note-moved listener failed:', cause)
+    }
+  }
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/lib/note-moves.ts:35]: Fixed in 339115c: both healing paths now take an onMoved hook (reconcileIndex via IndexPassOptions, watcher batches via subscribeIndexChanges), wired through createGraphIndex to followHealedMove — it carries any live session to the new path and emits noteMoved, so routes, history, and open panes follow an external rename exactly as an in-app one (and an open pane's next save can no longer resurrect the dead path). Pinned by core + desktop tests.

File: apps/desktop/src/lib/filename-migration.ts [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+      const target = await slugPathForTitle(candidate.path, parsed.title)
+      if (target !== candidate.path) {
+        await moveNoteCarryingSession(candidate.path, target, generation)
+      }
+      result.moved += 1
+    } catch (cause) {
+      result.failed.push({
+        path: candidate.path,
+        message: cause instanceof Error ? cause.message : String(cause),
+      })
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/lib/filename-migration.ts]: Fixed in 3069030: migrateUlidNotes uses errorMessage(cause) like the rest of the codebase, and the failure test now asserts the recorded message for an IPC-shaped { kind, message } error.

File: apps/desktop/src/editor/use-note-document.ts [hunk 10]
```diff
@@ -88,86 +99,57 @@ export function useNoteDocument(
...
-      // this session as "open" — its alias goes to disk (or to a reopened
-      // pane's live session, which registers under the same path).
-      unregisterDocument()
-      // Disposal flushes pending edits to the session's own path — the
-      // path-switch "final flush" lives here, not in cross-note bookkeeping.
-      // The flush's landed save reaches the tracker via onContent('saved');
-      // settle after it so a just-edited title still renames on the way out.
-      const settled = session.flush()
-      session.dispose()
-      if (coordinator) {
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/use-note-document.ts:154]: Fixed in 9355b0e: the watcher handler compares against the session's current path at event time (via the binding accessor) instead of the route prop, so upserts at the retargeted path reconcile inside the pre-rerender window. Pinned by a regression test that lands an external edit in exactly that window.

File: apps/desktop/src-tauri/src/db/write.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src-tauri/src/db/write.rs:164]: Superseded in d0fb455 (deliberate simplification): the same-id adoption this thread led to is removed — an occupied destination (db row or disk file) now always refuses the move, consistently across both layers. The race window is sub-millisecond after a 5s save-quiet period; a refused rename costs only cosmetic filename drift that the next settled rename retries, and nothing is ever deleted or overwritten.

File: apps/desktop/src/components/filename-migration-prompt.tsx [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+  if (
+    root === null ||
+    generation === null ||
+    declined ||
+    dismissed ||
+    candidates === null ||
+    candidates.length === 0
+  ) {
+    return null
+  }
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/components/filename-migration-prompt.tsx]: Working as designed, now documented (e0695de): the scenario can't occur — every path that re-triggers indexing goes through status 'opening', which unmounts the workspace and this dialog with it (state resets), so an open offer never coexists with a fresh reconcile. indexing deliberately gates the scan (count quality), not the render: hiding an offered dialog on a background flag would yank it from under the user's cursor, and writes during a reconcile are an everyday tolerated condition (autosaves) — the runner re-validates every candidate.

File: apps/desktop/src/editor/move-note.ts [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+  generation: number,
+): Promise<void> {
+  const owner = openSession(from)
+  if (owner !== null) {
+    await owner.flush()
+    owner.retarget(to)
+    retargetOpenDocument(from, to, owner)
+  }
+  try {
+    await moveNoteIndexed(from, to, generation)
```
[the remark this answers is not in the document]
Comment (maccman) [reply apps/desktop/src/editor/move-note.ts:39]: Working as designed, now documented (e0695de): this is the race the simplification commit (d0fb455) explicitly accepted — the window is one IPC round-trip behind a 5s save-quiet gate and an 800ms debounce, and if it ever fires the write lands at the destination (where the note is about to live), the move refuses, and at worst an orphan copy surfaces via the duplicate-id flag. The alternative ordering (retarget after the move) fails worse: the same race resurrects the OLD path holding the newest bytes. Rationale recorded in the plan's risk log and the moveNoteCarryingSession docstring.
